anitya/lib/librariesio_consumer.py

```diff
@@ -0,0 +1,205 @@
+        hostname = socket.gethostname().split('.', 1)[0]
+        fedmsg.init(name='anitya-libraryio.%s' % hostname)
+        fedmsg.meta.make_processors(**self.hub.config)
+
+    def consume(self, message):
+        """
+        This method is called when a message with a topic this consumer
+        subscribes to arrives.
+
+        Args:
+            message (dict): The fedmsg to process.
+        """
+        project = self._fedmsg_to_project(message)
+        if project is not None:
+            # Double check the validity of the update
+            if self._validate_update(project, message):
+                session = Session()
+                session.add(project)
+                session.commit()
+                Session.remove()
+
+    def _validate_update(self, project, message):
+        """
+        Validate the project was updated.
+
+        This serves two purposes. Firstly, we connect to libraries.io over HTTP
+        so we don't have any authentication. Secondly, we might map the
+        libraries.io project to the wrong Anitya project, so this is a good
+        opportunity to catch and log those problems.
+        """
+        librariesio_version = message['body']['msg']['data']['version']
+        try:
+            anitya_version = get_version(project)
+        except exceptions.AnityaPluginException as e:
+            # TODO what to do here? Bail? Trust libraries.io?
+            _log.warning('libraries.io found an update for %r, but we failed with %s',
+                         project, str(e))
+            return False
+
+        # TODO handle things like 'v' prefix. Should be a general version normalizer
+        # used everywhere.
+        # TODO handle when we find a different version than librariesio. Create the project anyway?
+        equivalent_versions = (librariesio_version == anitya_version)
+        if not equivalent_versions:
+            _log.info('libraries.io found %r had a latest version of %s, but Anitya found %s',
+                      project, librariesio_version, anitya_version)
+        return equivalent_versions
+
+    def _fedmsg_to_project(self, message):
+        """
+        Transform a librariesio fedmsg into an Anitya project
+
+        Args:
+            message (dict): The fedmsg received by :meth:`self.consume`
+
+        Returns:
+            Project: the Anitya project that maps to the librariesio project;
+                this may be a new project.
+            None: If there is no ecosystem that maps to the librariesio platform
+        """
+        librariesio_msg = message['body']['msg']['data']
+        name = librariesio_msg['name']
+        platform = librariesio_msg['platform']
+        try:
+            ecosystem = list(filter(
+                lambda eco: eco.librariesio_name == platform, ECOSYSTEM_PLUGINS.get_plugins()))[0]
+        except IndexError:
+            # TODO probably better to raise an exception here
+            return
+
+        session = Session()
```

Comment:
Based on my (not yet complete) understanding of SQLAlchemy, the whole point of using its ``scoped_session`` is to create a session registry where each thread gets its own session so you don't have to pass session objects around. The session "created" here is the same session "created" [here](https://github.com/release-monitoring/anitya/pull/433/files#diff-5379d281c0b6a0060c136e2c384cd573R133). When ``session.remove()`` is called the session is removed from the registry and its resources released. Next time ``Session()`` is called a new session is placed in the thread-local registry.

Not really on-topic, but we're not using a lot of nice SQLAlchemy features. I think in the near future we need to alter all the database APIs to use custom query classes (where appropriate) in combination with the [query_property](http://docs.sqlalchemy.org/en/latest/orm/contextual.html#sqlalchemy.orm.scoping.scoped_session.query_property) rather than all those static methods on the models that accept sessions. The backend should create a scoped session rather than relying on the Flask application do to so (otherwise we should be using flask-sqlalchemy since we're tying the backend to the frontend anyway).